### DISCUSSION TITLE
Pages i18n breaks Intercepting Routes

### DISCUSSION THREAD

**Discussion by Ayterx** [vercel/next.js#60592] [App Router]
### Link to the code that reproduces this issue or a replay of the bug

https://github.com/Ayterx/intercepting-i18n-bug

### To Reproduce

1. Create an intercept route, whether it is inside a Parallel Route or not.
2. Add i18n to next.config.js.
3. Soft navigate to the `Intercept Route`, and it will be ignored.

Live example: [intercept-i18n-bug.vercel.app](https://intercept-i18n-bug.vercel.app/)

### Current vs. Expected behavior

Intercepting Routes work as intended, and Pages i18n will not affect the App Router.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Home
    Binaries:
      Node: 18.16.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.20-canary.16
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Internationalization (i18n)

### Additional context

When `i18n` is included in next.config.js, intercept routes are completely ignored, and Next.js renders the default route.
`/app`
` /@modal/(.)notifications` (Will never render when soft navigation)
` /notifications` (Will render always)

In version 13.4.20-canary.12 or earlier, the app won't build, and an error occurs:
`Error occurred prerendering page "/dashboard/(.)notifications". Read more: https://nextjs.org/docs/messages/prerender-error
Error: Unsupported Server Component type: undefined`

This might be a bug related to `Parallel Routes`

**Comment by mlmetzner** [comment]
is your /@modal/  in app/ or is there a language folder [lang]/ , because afaik (.) only catches on the top level.

**Comment by Ayterx** [comment]
@mlmetzner It doesn't have [lang]. I tried adding [lang], but it didn't have any effect.

My app has a Pages Router with i18n, and the dashboard(App Router) doesn't have i18n. It was working fine until I tried adding  `intercept route`, where I found this bug with Pages i18n and `intercept route`.

**Comment by hadnet** [comment]
Same thing with me, it doesn't work using i18n enabled, i'm using the app router and the intercepting routes is completely ignored. Any solution?

**Comment by Jokcy** [comment]
Same issue here, spend me one day to debug this weird none documented issue

**Comment by finkrer** [comment]
Wow, I have just spent 2 work days looking for the reason only to find it's a known issue.

Here's an alternative reproduction:

1. Clone the https://github.com/vercel-labs/nextgram repository
2. Observe that it's working correctly
3. Add the following to `next.config.js`
```js
  i18n: {
    locales: ["default"],
    defaultLocale: "default",
  },
```
4. See that the intercept is now ignored and the normal route is opened instead

**Comment by samcx** [comment]
@Ayterx Thank you for submitting an issue!

Unfortunately, `i18n` does not work with App Router in general (it's not documented as an option in the App Router docs for `next.config.js` → https://nextjs.org/docs/app/api-reference/next-config-js).

You will need to use Middleware instead to set up Internationalization with App Router → https://nextjs.org/docs/app/building-your-application/routing/internationalization.

Since this not an unexpected bug with Parallel routes, I will be converting this to a Discussion.

**Reply by finkrer** [reply]
Hey @samcx. Just because it's unsupported doesn't mean having leftover configs from pages i18n should silently break unrelated app router features. The issue is not that the option doesn't work with the app router, it's that it breaks parallel routes. 

People have production Next.js apps that they will port to the new router and encounter this issue. The fact that this a legacy config option that can be safely removed (in fact, has to be removed because of the bug) will likely be overlooked. 

I think moving this to a discussion wasn't a good idea. This is an actual bug that needs fixing. I would suggest that first of all, the option is properly ignored by the app router, and second, there should be a warning that it is no longer necessary and can be removed.

**Reply by samcx** [reply]
@finkrer Thanks for sharing your thoughts! Definitely on the same page that not showing a warning (or safely ignoring) for the config when using App Router is valid. It's technically not legacy since it works with Pages Router, so we don't plan on removing it.

I'll create a new issue to better highlight the general issue (e.g., title: no warning (or safely ignoring) when using i18n with app router)!

**Reply by Ayterx** [reply]
My issue is repeatedly being misunderstood, so let's use an example for better clarity. Imagine your team has a marketing site using Pages Router and the website includes localization (i18n) through next.config.js. Now if your team aims to incrementally adopt App Router (adopting it piece by piece) I wish them good luck, because it's practically impossible. You'll encounter the strangest bugs; intercepting routes never work in production, and the app won't even build. There are numerous other issues related to this, and this particular issue has persisted for over a year.

If you want to use App Router, your team will likely have to rebuild the entire app with App Router from scratch. There's no such thing as 'incrementally adopting' App Router

**Reply by samcx** [reply]
@Ayterx it is possible to incrementally adopt App Router (without rebuilding it from scratch or breaking existing pages)—it was done incrementally for the Vercel dashboard, our docs, etc.

It's unfortunate that `i18n` in the `next.config.js` does not work with App Router, but that is the current state of it today until we we're able to land that or a better story around internationalization. In the meantime, it is possible to implement the same using Middleware.

- https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration

> intercepting routes never work in production, and the app won't even build

A lot of parallel & interception bug fixes have landed recently (still a few more to go), so I would recommend testing on the latest if you are still experiencing this!

As mentioned above, we need to still create a warning for this when using App Router.

**Reply by Ayterx** [reply]
> it was done incrementally for the Vercel dashboard, our docs, etc.

The Vercel site doesn't have i18n which makes it possible to incrementally adopt App Router, basically the problem happens when you have two app Pages (use i18n) and App Router.

**Reply by samcx** [reply]
@Ayterx this is true! Was just making a point about incrementing to App Router in general (outside of edge cases such as the `i18n` config).

You can mimic what's possible (and more since Middleware is technically more flexible) with `i18n` with Middleware. The downside would be you would be accruing Middleware usage, but that's typically a minor cost since it's quite cheap to run Middleware on every request (e.g., https://vercel.com/docs/functions/edge-middleware/usage-and-pricing).

 - https://nextjs.org/docs/app/building-your-application/routing/internationalization#routing-overview

**Reply by iamjohnlong** [reply]
I was looking into this today and ended up here. I have an app on `pages` and was looking to starting to incrementally adopt the `app` router and I'm scratching my head on how to to do it. My app is using sub domains for locales using https://nextjs.org/docs/pages/building-your-application/routing/internationalization#domain-routing . How is it possible to have both? I have a catch all path in my pages dir `pages/[productId]` I can't have a catch all for locales in `app/[locale]` the paths are conflicted.

**Reply by iamjohnlong** [reply]
@leerob thoughts on how I can work through this?

**Reply by corbin-mosher** [reply]
Honestly, my biggest disappointment here is the lack of a very obvious and thorough migration step for i18n within https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration. I would have expected dropping a core feature like i18n to warrant a thorough guide and warning.

There is only one hint in the guide in an unrelated routing hooks context.

> The locale, locales, defaultLocales, domainLocales values have been removed because built-in i18n Next.js features are no longer necessary in the app directory. [Learn more about i18n](https://nextjs.org/docs/app/building-your-application/routing/internationalization).

When creating migration plans and timelines it is easy to miss this hurdle.

> outside of edge cases such as the i18n config

It feels like the treatment of i18n as an `edge case` likely contributes to that. When in reality it is a hugely important feature for major sites.

I am worried that the middleware solution won't be a simple 1:1 replacement, but I will make the attempt and keep my fingers crossed 🤞

**Reply by samcx** [reply]
@corbin-mosher Thank you for the feedback! I do think Internationalization should be a core feature, but I also agree it needs to be done right in App Router. It's just something we haven't been able to prioritize further yet.

The reason for the lack of a migration is because what we found out is that folks tend to want to have their own implementation for Internationalization—there isn't a one-fit-all solution. You can technically do all the Internationalization routing you need through Middleware, hence we are recommending that for now. Maybe the best path in the future for Internationalization with App Router is a less-rigid solution (an in-between).

Let us know if you run into any difficulties using Middleware!

**Reply by corbin-mosher** [reply]
@samcx the biggest challenge with migrating to middleware from the page's directory `i18n` feature is the loss of the client side routing support for an implicit locale path param.

All of the usages of `useRouter().replace`, `useRouter().push`, and `next/link` now require prepending the present locale to the start of the destination path. Without it, navigations between one page's directory path to another will be considered unhandled paths and cause a full page refresh.

This get exasperated by the fact that large sites that use internationalization also are likely to have many urls provided from headless CMS data rather than directly in code. So you need to basically extend the Nextjs router and link components so that they handle forcing a locale when it is not present. Doable for sure, but leaves the code base in an awkward state where engineers need to avoid using the Next features directly. Which then requires making some linting rules.

Bumpy road but possible 😓 

Let me know though if you think I am missing something that would make things easier.

**Edit:** I forgot to mention the fact that its also possible for a domain to be powered by more than just a Nextjs application. So relative links that move from the Nextjs application to pages handled by other applications are much much more difficult to deal with. Since they may not support a locale path param, but knowing not to inject one is not trivial always.

**Reply by samcx** [reply]
@corbin-mosher Thank you for the feedback!

I don't believe you're missing anything—this is expected (no missing config that could easy things like `locale` and `<Link>`.

Just to re-iterate, part of the decision for not having a dedicated `i18n` feature/config in App Router (yet) is because the feedback we've received was that folks were feeling the bumpiness trying to get their own Internationalization needs going with it. :tldr:, it was too rigid for some folks needs.

At least with Middleware, it's very customizable, which also means you have to wire about more things to get things working.

Nonetheless, I still believe we could provide something better than what was provided in Pages Router!

**Comment by joaogarin** [comment]
> Unfortunately, i18n does not work with App Router in general (it's not documented as an option in the App Router docs for next.config.js → https://nextjs.org/docs/app/api-reference/next-config-js).

How is one supposed to incrementally adopt app router if something like i18n (whch is needed on pages) does not work on the app directory?

> Honestly, my biggest disappointment here is the lack of a very obvious and thorough migration step for i18n within https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration. I would have expected dropping a core feature like i18n to warrant a thorough guide and warning.

This, couldn't agree more. Especially given how the i18n feature was presneted as such an important feature in next conf a couple years ago.

**Comment by corbin-mosher** [comment]
I just wanted to share how I was able to go about migrating off of Next's i18n in a way that minimized the blast radius of the effort. It is different than the provided guidance and might be useful to others to escape this discontinued feature. The provided guidance required huge blast radius with too many edge cases. nesting directories and overriding `next/link` etc... not the best. Here is what I did:

## Avoid messy folder madness

We did not nest your `app` and `pages` directory contents within a `[locale]` directory. Instead of taking a mindset that client-side routing had to support `i18n`, we went the other direction. Instead I configured the middleware to STRIP i18n path params if present. Not add them. This protects us from ever having to server-side render a path that contains localization information.

## Set cookies

In the middleware that strips i18n off of the request's route, I go ahead and set that stripped info into cookies. If there is no i18n info in the URL, the cookies are either respected if pre-existing, or we make default decisions based on our CDN's cookie for their requesting country and their lang preferences.

## Read the cookies in RSCs

For internationalizing choices within RSCs, we can just read the cookies that we have set in middleware. Based on those we do a dynamic import of the correct message/translation files wrapped in React's `cache` function to avoid unnecessary repeated work across RSCs.

## Use React context for SSR and Client/Hydration

Expose the cookies to your code with React context. We read the cookies in an RSC, convert them to a header string, and pass them to a client side context which converts them to an object that is easy to leverage/read. Just be careful not to pipe HTTP only cookies into client side context if you use them.
